File: lib/ts/src/internal/binary/ClientBinaryStrategy.ts
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

import { BinaryEncodingCache } from "./BinaryEncodingCache.js";

export class Checksum {
    constructor(
        public value: number,
        public expiration: number,
    ) {}
}

export class ClientBinaryStrategy {
    private primary: Checksum | null = null;
    private secondary: Checksum | null = null;
    private lastUpdate: Date = new Date();
    private binaryEncodingCache: BinaryEncodingCache;

    constructor(binaryEncodingCache: BinaryEncodingCache) {
        this.binaryEncodingCache = binaryEncodingCache;

        const initialChecksums = binaryEncodingCache.getChecksums();

        const randomPrimary = initialChecksums[0];
        const randomSecondary = initialChecksums[1];

        if (randomPrimary) {
            this.primary = new Checksum(randomPrimary, 0);
        }

        if (randomSecondary) {
            this.secondary = new Checksum(randomSecondary, 0);
        }
    }

    updateChecksum(newChecksum: number): void {
        if (!this.primary) {
            this.primary = new Checksum(newChecksum, 0);
            return;
        }

        if (this.primary.value !== newChecksum) {
            let expiredChecksum = this.secondary;
            this.secondary = this.primary;
            this.primary = new Checksum(newChecksum, 0);
            if (this.secondary) {
                this.secondary.expiration += 1;
            }

            if (expiredChecksum) {
                this.binaryEncodingCache.remove(expiredChecksum.value);
            }

            return;
        }

        this.lastUpdate = new Date();
    }

    getCurrentChecksums(): number[] {
        if (!this.primary) {
            return [];
        } else if (!this.secondary) {
            return [this.primary.value];
        } else {
            const minutesSinceLastUpdate = (Date.now() - this.lastUpdate.getTime()) / (1000 * 60);

            // Every 10 minute interval of non-use is a penalty point
            const penalty = Math.floor(minutesSinceLastUpdate / 10) + 1;

            if (this.secondary) {
                this.secondary.expiration += 1 * penalty;
            }

            if (this.secondary && this.secondary.expiration > 5) {
                this.secondary = null;
                return [this.primary.value];
            } else {
                return [this.primary.value, this.secondary.value];
            }
        }
    }
}

File: lib/go/internal/binary/ClientBinaryStrategy.go
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

package binary

import (
	"sync"
	"time"
)

type checksumRecord struct {
	value      int
	expiration int
}

// ClientBinaryStrategy tracks checksum rotation for client binary encoding.
type ClientBinaryStrategy struct {
	cache      BinaryEncodingCache
	primary    *checksumRecord
	secondary  *checksumRecord
	lastUpdate time.Time
	mu         sync.Mutex
}

// NewClientBinaryStrategy constructs a ClientBinaryStrategy instance.
func NewClientBinaryStrategy(cache BinaryEncodingCache) *ClientBinaryStrategy {
	return &ClientBinaryStrategy{
		cache:      cache,
		lastUpdate: time.Now(),
	}
}

// UpdateChecksum records a newly observed checksum, rotating the primary/secondary slots when necessary.
func (s *ClientBinaryStrategy) UpdateChecksum(newChecksum int) {
	s.mu.Lock()
	defer s.mu.Unlock()

	now := time.Now()

	if s.primary == nil {
		s.primary = &checksumRecord{value: newChecksum}
		s.lastUpdate = now
		return
	}

	if s.primary.value != newChecksum {
		expired := s.secondary
		s.secondary = s.primary
		s.primary = &checksumRecord{value: newChecksum}
		if s.secondary != nil {
			s.secondary.expiration++
		}
		if expired != nil {
			s.cache.Remove(expired.value)
		}
		s.lastUpdate = now
		return
	}

	s.lastUpdate = now
}

// GetCurrentChecksums returns the current checksum order, retiring the secondary checksum after inactivity.
func (s *ClientBinaryStrategy) GetCurrentChecksums() []int {
	s.mu.Lock()
	defer s.mu.Unlock()

	if s.primary == nil {
		return []int{}
	}

	if s.secondary == nil {
		return []int{s.primary.value}
	}

	minutesSinceUpdate := time.Since(s.lastUpdate).Minutes()
	penalty := int(minutesSinceUpdate/10.0) + 1
	if penalty < 1 {
		penalty = 1
	}

	s.secondary.expiration += penalty
	if s.secondary.expiration > 5 {
		s.secondary = nil
		return []int{s.primary.value}
	}

	return []int{s.primary.value, s.secondary.value}
}
